packages/client/src/upload.ts

```diff
@@ -212,9 +221,7 @@ export async function* upload(
 
   while (Object.keys(uploadList).length > 0) {
     try {
-      const event = await Promise.race(
-        Object.keys(uploadList).map((key): Promise<any> => uploadList[key])
-      );
+      const event = await Promise.race(Object.values(uploadList));
```

Comment:
**question (non-blocking):** ‏Unrelated, but is `Promise.race` correct here?

This will make the first settled promise be the result of the race. If a following promise rejects, this logic will not handle that. This seems wrong. Maybe we visit this in a follow up PR.

Reply:
My understanding is that since this is in a while loop, it handles every promise in the order they finish. It would `yield` an initial success, then `return yield` a following rejection.

Though I'm definitely not used to all this async logic + semaphore + generator function behavior and I'm not sure I understand it all, so could be wrong.

Reply:
Oh, it's finding the first settled one and removing it. This makes sense.